Mathlib/Tactic/Linter/Header.lean

```diff
@@ -0,0 +1,298 @@
+        s!"In the past, importing 'Lake' in mathlib has led to dramatic slow-downs of the linter \
+          (see e.g. mathlib4#13779). Please consider carefully if this import is useful and \
+          make sure to benchmark it. If this is fine, feel free to allow this linter.")
+  return output
+
+/-- Check the syntax `imports` for syntactically duplicate imports.
+The output is an array of `Syntax` atoms whose ranges are the import statements,
+and the embedded strings are the error message of the linter.
+-/
+def duplicateImportsCheck (imports : Array Syntax)  : Array Syntax := Id.run do
+  let mut output := #[]
+  let mut importsSoFar := #[]
+  for i in imports do
+    if importsSoFar.contains i then
+      output := output.push (mkAtomFrom i s!"Duplicate imports: {i} already imported")
+    else importsSoFar := importsSoFar.push i
+  return output
+
+/--
+The "header" style linter checks that a file starts with
+```
+/-
+Copyright ...
+Apache ...
+Authors ...
+-/
+
+import*
+module doc-string*
+rest
+```
+It emits a warning if
+* the copyright statement is malformed;
+* `Mathlib.Tactic` is imported;
+* any import in `Lake` is present;
+* the first non-`import` command is not a module doc-string.
+
+The linter allows `import`-only files and does not require a copyright statement in `Mathlib.Init`.
+-/
+register_option linter.style.header : Bool := {
+  defValue := false
+  descr := "enable the header style linter"
+}
+
+namespace Style.header
+
+@[inherit_doc Mathlib.Linter.linter.style.header]
+def headerLinter : Linter where run := withSetOptionIn fun stx ↦ do
+  unless Linter.getLinterValue linter.style.header (← getOptions) do
+    return
+  if (← get).messages.hasErrors then
+    return
+  let mainModule ← getMainModule
+  if mainModule == `Mathlib then return
+  let fm ← getFileMap
+  let md := (getMainModuleDoc (← getEnv)).toArray
+  -- the end of the first module doc, or the end of the file if there are no module docs.
+  let firstDocModPos := match md[0]? with
+                          | none     => fm.positions.back
+                          | some doc => fm.ofPosition doc.declarationRange.endPos
+  unless stx.getTailPos? == some firstDocModPos do
+    return
+  -- We try to parse the file up to `firstDocModPos`.
+  let upToStx ← parseUpToHere firstDocModPos <|> (do
+    -- If parsing failed, there are some non-module docs, so we parse until the end of the
+    -- imports, adding an extra `section` after, so that we trigger a "no module doc" warning.
+    let fil ← getFileName
+    let (stx, _) ← Parser.parseHeader { input := fm.source, fileName := fil, fileMap := fm }
+    parseUpToHere (stx.getTailPos?.getD default) "\nsection")
+  let importIds := getImportIds upToStx
+  -- Report on broad or duplicate imports.
+  for (imp, msg) in broadImportsCheck importIds do
+    Linter.logLint linter.style.header imp msg
+  for out in duplicateImportsCheck importIds do
+    let .atom _ errorMessage := out
+     | return
```

Comment:
Are you worried here that `getImportIds` will not return an `Array` of `.atom`s?  Also, I think that using `continue` (untested) *might* work better since it may not prevent the rest of the array to be parsed, even if one entry is not an `.atom`.
```suggestion
    let .atom _ errorMessage := out
     | continue
```

Reply:
That is a reasonable point.
I am not worried about anything here: the point is merely that `out` only *returns** an `.atom`; the other case is simply unreachable. I've added a comment.